robotiq_joint_state_publisher/src/s_model_joint_states.cpp

```diff
@@ -0,0 +1,208 @@
+ */
+class Finger {
+ public:
+  Finger() { position = 0; }				///< Default constructor for creating Finger, position is set 0
+  Finger(int pos) { position = pos; }			///< Create Finger, position is @param pos
+  Finger(const Finger &f) { position = f.position; }	///< Create Finger, position is taken form @param Finger
+  inline double joint_1() const;			///< joint_1 value for Finger
+  inline double joint_2() const;			///< joint_2 value for Finger
+  inline double joint_3() const;			///< joint_3 value for Finger
+  int position;						///< Position of the Finger
+};
+
+/**
+ * Calculate joint_1 based on the Finger position.
+ */
+inline double Finger::joint_1() const {
+  if(0 <= position && position <= 140) return (70.0/148.0 * DEG_TO_RAD) * position;
+  else return 70.0 * DEG_TO_RAD;
+}
+
+/**
+ * Calculate joint_2 based on the Finger position.
+ */
+inline double Finger::joint_2() const {
+  if(0 <= position && position <= 140) return 0.0;
+  else if(140 < position && position <= 240) return (90.0/100.0 * DEG_TO_RAD) * (position-140.0);
+  else return 90.0 * DEG_TO_RAD;
+}
+
+/**
+ * Calculate joint_3 based on the Finger position.
+ */
+inline double Finger::joint_3() const {
+  if(0 <= position && position <= 140) return (-70.0/140.0 * DEG_TO_RAD) * position;
+  else return -55.0 * DEG_TO_RAD;
+}
+
+/**
+ * Robotiq S-model with three Fingers.
+ */
+class Robotiq3 {
+ public:
+  Robotiq3() { mode = BASIC; scissor = 137; joint_positions.resize(11, 0.0); }		///< Default constructor for creating Robotiq3
+  inline double scissor_joint() const;							///< Joint values for so-called scissor joints
+  void callback(const robotiq_s_model_control::SModel_robot_input::ConstPtr &msg);	///< Callback function for "SModelRobotInput" topic
+  Finger finger_left;									///< Robotiq FINGER A
+  Finger finger_right;									///< Robotiq FINGER B
+  Finger finger_middle;									///< Robotiq FINGER C
+  int scissor;										///< Scissor position
+  int mode;										///< Robotiq grasping mode
+  std::vector<double> joint_positions;							///< All the joint values
+};
+
+/**
+ * Calculate scissor_joint based on the scissor position.
+ */
+inline double Robotiq3::scissor_joint() const {
+  // fully open is +20, fully closed is -10
+  // std::cout << scissor << std::endl;
+  if(0 <= scissor && scissor <= 15) return 20.0*DEG_TO_RAD;
+  else if(15 < scissor && scissor <= 240) return (20.0-30.0*((scissor-15.0)/225.0))*DEG_TO_RAD;
+  else return -10.0*DEG_TO_RAD;
+}
+
+/**
+ * Callback function for "SModelRobotInput" topic.
+ * */
+void Robotiq3::callback(const robotiq_s_model_control::SModel_robot_input::ConstPtr &msg) {
+  finger_left = Finger(msg->gPOA);		// set left finger position
+  finger_right = Finger(msg->gPOB);		// set right finger position
+  finger_middle = Finger(msg->gPOC);		// set middle finger position
+  scissor = msg->gPOS;				// set scissor position
+  mode = msg->gMOD;				// set mode
+
+  // if (mode == BASIC)        scissor = 137;
```

Comment:
Should this code be commented out?  If so, are the mode and constant values required?